### PR TITLE
Alerting: GitLab alert notifier

### PR DESCRIPTION
**Pull request by seanarnold**
**What this PR does / why we need it**:

This adds GitLab as a notification channel. 

Specifically, it allows the ability to natively send alerts to your GitLab project's [Alert HTTP endpoint](https://docs.gitlab.com/ee/operations/incident_management/integrations.html#single-http-endpoint) which will create an [Alert](https://docs.gitlab.com/ee/operations/incident_management/alerts.html) inside GitLab as part of the GitLab Incident Management product offering.

Features:
- Automatic resolution of GitLab alerts
- Tags are added to the alert payload

GitLab issue: https://gitlab.com/gitlab-org/gitlab/-/issues/334351
Video walkthrough: https://www.youtube.com/watch?v=S9-92Zu_HSg

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/discussions/37775

**Special notes for your reviewer**:
We plan on adding a Grafana 8 contact point as a fast-follow too.

### CODE REVIEW

**Comment by marefr** [comment]
Thank you for contributing to Grafana! However, we cannot accept this and will close it for now since the discussion #37775 referenced has no traction/activity/comments besides yours. Remember closing a PR doesn't mean we can reopen it in the future. I do believe that you can integrate with the [webhook notifier](https://grafana.com/docs/grafana/latest/alerting/old-alerting/notifications/#webhook), implementing a service that would receive Grafana webhook and forward to GitLab as a workaround.

We would signal accepting a discussion as a feature request by converting the discussion to an issue. Please continue the discussions in #37775. Thanks again.

File: docs/sources/alerting/old-alerting/notifications.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ Name | Type | Supports images | Support alert rule tags
 [DingDing](#dingdingdingtalk) | `dingding` | yes, external only | no
 [Discord](#discord) | `discord` | yes | no
 [Email](#email) | `email` | yes | no
+[GitLab](#google-hangouts-chat) | `gitlab` | no | yes
```
Comment (dnsmichi) [inline docs/sources/alerting/old-alerting/notifications.md:58]: @seanarnold The anchor should be `#gitlab`.

Comment (seanarnold) [reply docs/sources/alerting/old-alerting/notifications.md:58]: Thanks, will update.

**Comment by AUtech-dev** [comment]
@seanarnold Thanks for providing this PR. I have been locking for such a solution for our systems.

I have tested this PR in combination with a local GitLab instance (v14.3).

Creating alerts works fine. However, automatically closing an alert if the alert in Grafana goes away did not work out of the box. GitLab opened a new alert for the closing alert request and marked this one as closed immediately. Some research showed that the `start_time` for the end alert was different from the timestamp from the begin-alert.

From my understanding `evalContext *alerting.EvalContext.StartTime` and `evalContext *alerting.EvalContext.EndTime` in `submitAlert()` do not stand for the start and end time of the alert but only for the timestamps for the handling of this alert. This means that the `StartTime` during handling the end alert is roughly the current timestamp and not the time of the start of alert.

I changed the source to solve this problem:
- send either `start_time` or `end_time` but not both https://github.com/AUtech-dev/grafana/commit/877016d3d058a56363abc28f341f335bc47d9fb0
- add a `fingerprint` key to allow GitLab matching the both requests to the same alert https://github.com/AUtech-dev/grafana/commit/76b3282599c1086bb09105f1f13999fe06b08ba0
  -   the fingerprint creation is copied from the Pagerduty notifier

With this changes even auto ending an alert works for me.

You can find my commits here: https://github.com/AUtech-dev/grafana/commits/gitlab-notification-fix-end

Feel free to cherry pick the commits into this PR.

**Comment by seanarnold** [comment]
Thanks @AUtech-dev. I tested this and it worked fine so perhaps our scenarios are a little different. Regardless I'll see if I can reproduce and incorporate your patches next week 👍 .

In the meantime It'd be great if you could show some interest on this feature in the discussion (https://github.com/grafana/grafana/discussions/37775), as Grafana now requires enough perceived interest before accepting new features. So anything you can post in that discussion on how you'd like to use this feature would be appreciated 🙇 ❤️.

**Pull request closed without merging**